pdns/iputils.hh

```diff
@@ -369,6 +370,313 @@ private:
+    TreeNode* node = root.get();
+    node_type* value = nullptr;
+
+    if (key.getNetwork().sin4.sin_family == AF_INET) {
+      std::bitset<32> addr(be32toh(key.getNetwork().sin4.sin_addr.s_addr));
+      int bits = 0;
+      // we turn left on 0 and right on 1
+      while(bits < key.getBits()) {
+        uint8_t val = addr[31-bits];
+        if (val)
+          node = node->make_right();
+        else
+          node = node->make_left();
+        bits++;
+      }
+      // only create node if not yet assigned
+      if (!node->node4) node->node4 = unique_ptr<node_type>(new node_type());
+      value = node->node4.get();
+    } else {
+      uint64_t* addr = (uint64_t*)key.getNetwork().sin6.sin6_addr.s6_addr;
+      std::bitset<64> addr_low(be64toh(addr[1]));
+      std::bitset<64> addr_high(be64toh(addr[0]));
+      int bits = 0;
+      while(bits < key.getBits()) {
+        uint8_t val;
+        // we use high address until we are
+        if (bits < 64) val = addr_high[63-bits];
+        // past 64 bits, and start using low address
+        else val = addr_low[127-bits];
+
+        // we turn left on 0 and right on 1
+        if (val)
+          node = node->make_right();
+        else
+          node = node->make_left();
+        bits++;
+      }
+      // only create node if not yet assigned
+      if (!node->node6) node->node6 = unique_ptr<node_type>(new node_type());
+      value = node->node6.get();
+    }
+    // assign key
+    value->first = key;
+    // keep a copy of all nodes, handy later.
+    _nodes.push_back(value);
+
+    return *value;
+  }
+
+  void insert_or_assign(const key_type& mask, const value_type& value) {
+    insert(mask).second = value;
+  }
+
+  const node_type& at(const key_type& value) const {
+    const node_type* node = lookup(value);
+    if (node == nullptr) throw std::range_error(value.toString() + string(" not found"));
+    return *node;
+  }
+
+  const node_type* lookup(const key_type& value) const {
+    return lookup(value.getNetwork(), value.getBits());
+  }
+
+  const node_type* lookup(const ComboAddress& value, int max_bits = 128) const {
+    if (!root) return nullptr;
+
+    TreeNode *node = root.get();
+    node_type *ret = nullptr;
+
+    // exact same thing as above, except
+    if (value.sin4.sin_family == AF_INET) {
+      max_bits = std::max(0,std::min(max_bits,32));
+      std::bitset<32> addr(be32toh(value.sin4.sin_addr.s_addr));
+      int bits = 0;
+
+      while(bits < max_bits) {
+        // ...we keep track of last non-empty node
+        if (node->node4) ret = node->node4.get();
+        uint8_t val = addr[31-bits];
+        // ...and we don't create left/right hand
+        if (val) {
+          if (node->right) node = node->right.get();
+          // ..and we break when road ends
+          else break;
+        } else {
+          if (node->left) node = node->left.get();
+          else break;
+        }
+        bits++;
+      }
+      if (node->node4) ret = node->node4.get();
```

Comment:
maybe document that this line is only needed if `max_bits == 0` as otherwise the same line in the loop was already run for the current `node` value.